### ISSUE TITLE
Flaky DecodeSpec example

### ISSUE THREAD

**Issue by rossabaker** [http4s/http4s#4561]
```
==> X org.http4s.DecodeSpec.decode should either succeed or raise a CharacterCodingException  0.069s munit.FailException: /home/runner/work/http4s/http4s/tests/src/test/scala/org/http4s/DecodeSpec.scala:143
142:
143:  test("decode should either succeed or raise a CharacterCodingException") {
144:    forAll { (bs: Array[Byte], cs: Charset) =>

Failing seed: IBL0omGp6npzsX0tfYN5gMyayxqYsYysSeCU5hAY0dH=
You can reproduce this failure by adding the following override to your suite:

  override val scalaCheckInitialSeed = "IBL0omGp6npzsX0tfYN5gMyayxqYsYysSeCU5hAY0dH="

Falsified after 31 passed tests.
> ARG_0: [B@69b69ad1
> ARG_1: x-COMPOUND_TEXT
> ARG_0_ORIGINAL: [B@213c7441
```

**Comment by phongngtuan** [comment]
So this one only fails on jdk8, because the Charset `x-COMPOUND_TEXT` is not available on jdk11
```
java.lang.IllegalArgumentException: Non-positive length
	at java.nio.charset.CoderResult$Cache.get(CoderResult.java:202)
	at java.nio.charset.CoderResult$Cache.access$200(CoderResult.java:194)
	at java.nio.charset.CoderResult.malformedForLength(CoderResult.java:236)
	at sun.nio.cs.ext.COMPOUND_TEXT_Decoder.implFlush(COMPOUND_TEXT_Decoder.java:654)
	at java.nio.charset.CharsetDecoder.flush(CharsetDecoder.java:668)
	at org.http4s.internal.package$.flushLoop$1(package.scala:216)
	at org.http4s.internal.package$.$anonfun$decode$3(package.scala:222)
```

this can be reproduced by running with the string generated by seed
```scala
    val bs = Array[Byte](0,-100,-97,-126,-128,-1,1,-2,0,-16,-1,80,-1,0,36,-101)
```

but it could be reduced too just
```scala
    val bs = Array[Byte](70, 79, 79, -101) // "FOO" and then decimal -101
```

the crux of this is calling `decoder` on `sun.nio.cs.ext.COMPOUND_TEXT_Decoder` throws exception when it's in the middle of parsing. See [here](https://code.yawk.at/java/8/sun/nio/cs/ext/COMPOUND_TEXT_Decoder.java#652)

in this case, `state = CONTROL_SEQUENCE_PIF;` (https://code.yawk.at/java/8/sun/nio/cs/ext/COMPOUND_TEXT_Decoder.java#200) because the last byte in input stream is 0x9B = which is [09/11 | CSI | CONTROL SEQUENCE INTRODUCER](https://www.x.org/releases/X11R7.6/doc/xorg-docs/specs/CTEXT/ctext.html#Control_Characters)

the problem is that this internal `state` is not exposed so we can't tell if there is decoder is in "pending" state, although `result` is still underflow. (I guess) the only way to tell is to call flush and swallowing exception. or we can throw it, but then the property needs to be changed so that the bytes it randomly generates are valid, or we change the test to not fail when there are exceptions

i.e.
```
               case _ if result.isUnderflow =>
                 def flushLoop: Pull[F, String, Unit] =
-                  decoder.flush(charBuffer) match {
-                    case result if result.isUnderflow =>
-                      out
-                    case result if result.isOverflow =>
+                  Try(decoder.flush(charBuffer)) match {
+                    case Success(result) if result.isOverflow =>
                       out >> flushLoop
+                    case _ =>
+                      out
                   }
                 flushLoop.as(None)
```

Either way it's pretty simple so If you're okay then I can make the change :)

**Comment by rossabaker** [comment]
Wow. Impressive research!

I don't like just swallowing the exception in the decoder.  And clearly decoding can fail in more ways than the `CharacterCodingException`.  How about this:

1. Call `cs.decode(ByteBuffer.wrap(bs))`.
2. If it succeeds, run through our `decode` and `foldMonoid`.  Compare our String to the contents of the `CharBuffer` from step 1.
3. If it fails, run our `decode` and assert that it also fails.  Exception equality is hard, so don't be picky about the exception type.

So basically the property is that decoding the stream gives a consistent result with decoding directly on the charset.  

What do you think?

**Comment by phongngtuan** [comment]
sounds great! Let me give it a try

**Comment by phongngtuan** [comment]
so that fixes the "x-COMPOUND-TEXT" but looks like I ran into case where our decoding of utf8 is not consistent with plain nio (which gives the same result as fs2's `utf8Decode`).

for input below, ours throws exception while the others turns it into [Replacement Character](https://en.wikipedia.org/wiki/Specials_(Unicode_block)#Replacement_character)
```scala
val bs: Array[Byte] = Array[Byte](33, -6, -52) // 0x21 0xFA 0xCC
```
